src/main/java/com/google/devtools/build/lib/exec/CompactSpawnLogContext.java

```diff
@@ -471,83 +523,100 @@ private int logDirectory(
+   * <p>Since runfiles trees can be very large and, for tests, are only used by a single spawn, we
+   * store them in the log as a special entry that references the nested set of artifacts instead of
+   * as a flat directory.
    *
-   * @param root the path to the runfiles directory
-   * @param mapping a map from runfiles-relative path to the underlying artifact, or null for an
-   *     empty file
-   * @return the entry ID of the {@link ExecLogEntry.Directory} describing the directory.
+   * @param shared whether this runfiles tree is likely to be contained in more than one Spawn's
+   *     inputs
+   * @return the entry ID of the {@link ExecLogEntry.RunfilesTree} describing the directory.
    */
-  private int logRunfilesDirectory(
-      PathFragment root,
-      Map<PathFragment, Artifact> mapping,
+  private int logRunfilesTree(
+      RunfilesTree runfilesTree,
       InputMetadataProvider inputMetadataProvider,
-      FileSystem fileSystem)
+      FileSystem fileSystem,
+      boolean shared)
       throws IOException, InterruptedException {
     return logEntry(
-        root.getPathString(),
+        shared ? runfilesTree.getExecPath().getPathString() : null,
         () -> {
-          ExecLogEntry.Directory.Builder builder =
-              ExecLogEntry.Directory.newBuilder().setPath(root.getPathString());
-
-          for (Map.Entry<PathFragment, Artifact> entry : mapping.entrySet()) {
-            String runfilesPath = entry.getKey().getPathString();
-            Artifact input = entry.getValue();
-
-            if (input == null) {
-              // Empty file.
-              builder.addFiles(ExecLogEntry.File.newBuilder().setPath(runfilesPath));
-              continue;
-            }
-
-            Path path = fileSystem.getPath(execRoot.getRelative(input.getExecPath()));
-
-            if (isInputDirectory(input, path, inputMetadataProvider)) {
-              builder.addAllFiles(expandDirectory(path, runfilesPath, inputMetadataProvider));
-              continue;
-            }
-
-            Digest digest =
-                computeDigest(
-                    input,
-                    path,
-                    inputMetadataProvider,
-                    xattrProvider,
-                    digestHashFunction,
-                    /* includeHashFunctionName= */ false);
-
-            builder.addFiles(
-                ExecLogEntry.File.newBuilder().setPath(runfilesPath).setDigest(digest));
+          Preconditions.checkState(workspaceName.equals(runfilesTree.getWorkspaceName()));
+
+          ExecLogEntry.RunfilesTree.Builder builder =
+              ExecLogEntry.RunfilesTree.newBuilder()
+                  .setPath(runfilesTree.getExecPath().getPathString())
+                  .setLegacyExternalRunfiles(runfilesTree.isLegacyExternalRunfiles());
+
+          //
+          // The logic below together with the rule that artifacts override symlinks results in the
+          // following order of precedence:
```

Comment:
Should this comment be moved to the reconstructor? The way I see it, the proto itself doesn't enforce a precedence, it's up to the reconstructor to visit fields in the correct order.

Reply:
It's actually outdated now that we don't preresolve the symlinks, so I just removed it. The `spawn.proto` file itself has a similar comment that's up to date.